caffe2/operators/top_k.cc

```diff
@@ -26,147 +33,184 @@ template <typename T>
 struct ValueCmp {
   bool operator()(
       const std::pair<T, TIndex>& lhs,
-      const std::pair<T, TIndex>& rhs) {
-    return (
-        lhs.first > rhs.first ||
-        (lhs.first == rhs.first && lhs.second < rhs.second));
+      const std::pair<T, TIndex>& rhs) const {
+    return lhs.first > rhs.first ||
+        (lhs.first == rhs.first && lhs.second < rhs.second);
   }
 };
 
-// Define these two names to allow lookup into the 2d tensors like
-// mytensor(i, j)
 template <typename T>
-using EigenMatrixMapRowMajor = Eigen::Map<
-    Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+void GetTopK(
+    const T* input,
+    const TIndex n,
+    const int k,
+    const TIndex src_offset,
+    const TIndex dst_offset,
+    const TIndex stride,
+    T* values,
+    TIndex* indices,
+    TIndex* flatten_indices) {
+  const T* src_ptr = input + src_offset;
+  std::vector<std::pair<T, TIndex>> vec;
+  vec.reserve(k);
+  for (TIndex i = 0; i < k; ++i) {
+    vec.emplace_back(*src_ptr, i);
+    src_ptr += stride;
+  }
+  std::priority_queue<
+      std::pair<T, TIndex>,
+      std::vector<std::pair<T, TIndex>>,
+      ValueCmp<T>>
+      pq(ValueCmp<T>(), std::move(vec));
+  for (TIndex i = k; i < n; ++i) {
+    if (pq.top().first < *src_ptr) {
+      pq.pop();
+      pq.emplace(*src_ptr, i);
+    }
+    src_ptr += stride;
+  }
+  TIndex dst_pos = dst_offset + (k - 1) * stride;
+  while (!pq.empty()) {
+    const auto& item = pq.top();
+    values[dst_pos] = item.first;
+    indices[dst_pos] = item.second;
+    if (flatten_indices != nullptr) {
+      flatten_indices[dst_pos] = src_offset + item.second * stride;
+    }
+    pq.pop();
+    dst_pos -= stride;
+  }
+}
 
 template <typename T>
-using ConstEigenMatrixMapRowMajor = Eigen::Map<
-    const Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>>;
+void SetTopKGradient(
```

Comment:
Nit: briefly add some simple comment about the core algorithm/formula. It will be very helpful to understand how we compute the gradient.